bounswe/bounswe2017group3
# Initial configuration for our events application and basic user endpoints

## About

* **Related Issues:** https://github.com/bounswe/bounswe2017group3/issues/62
* **Other Related Links:**
  - Spring documentation: https://docs.spring.io/spring/docs/current/spring-framework-reference/html/
  - PostgreSQL documentation: https://www.postgresql.org/docs/
  - LiquiBase documentation: http://www.liquibase.org/documentation/index.html
  - Writing tests for Spring applications: http://www.lucassaldanha.com/unit-and-integration-tests-in-spring-boot/
  - What is Heroku?: https://www.heroku.com/what
  - What is a database migration?: http://www.iheavy.com/2011/06/15/database-migration-what-is-it-and-why-is-it-important/

Someone had to do the heavy work and create an application from scratch. I've created and configured our events application REST API using **Spring** framework (which we've agreed on), configured our deployment environment using **Heroku**, configured our database using **PostgreSQL** and **LiquiBase** and created the basic user endpoints which are `GET /user` and `POST /user`. I think it'll be easier for others to continue since they now have a code example.

Currently, tests and documentation of our API are missing. I think we can finish them after creating a skeleton application which works as expected 😄

Also, note that after adding the `Event` model, we have to implement a many-to-one dependency between `User` and `Event`s 👍 

### Changes

- Created a basic Spring Boot application.
- Made the necessary configurations in the application to run it as a REST API instead of a full MVC application.
- Configured our deployment environment in Heroku. **You can see our application live on https://bounswe2017group3.herokuapp.com/.**
- Decided to use PostgreSQL as our database and made its configurations.
- Decided to use LiquiBase for the database migrations and made the first migration to create a `User` model. I strongly suggest you to check out the documentation of LiquiBase from here: http://www.liquibase.org/documentation/index.html . If you don't know what a database migration is, please read this too: http://www.iheavy.com/2011/06/15/database-migration-what-is-it-and-why-is-it-important/
- Configured our Spring application so that adding `@ResponseBody` annotation to a method makes it return JSON. Example:
```java
@RequestMapping(method = RequestMethod.GET, produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
    public @ResponseBody List<User> user(ModelMap model) {
        return repository.findAll();
    }
```
- Created 2 endpoints for `User`:
  - `GET /user`: Returns all the users in an array.
  - `POST /user`: Creates a user from the given form data parameters.
- Implemented a basic encryption method to store (it's implicitly done in the `User` class) and check (by `checkPassword` method on `User`) user passwords safely.

## Visuals

Fields and methods of `User` model:
<img width="561" alt="screen shot 2017-04-16 at 00 27 00" src="https://cloud.githubusercontent.com/assets/13895224/25066876/75148ff0-223b-11e7-8312-78523e943d02.png">

Example queries made on Postman:
![screen shot 2017-04-15 at 23 08 36](https://cloud.githubusercontent.com/assets/13895224/25066843/bc18e49c-223a-11e7-962a-f2085d722ed0.png)
![screen shot 2017-04-15 at 23 08 39](https://cloud.githubusercontent.com/assets/13895224/25066844/bc1ddb82-223a-11e7-99ed-f9b1a532967f.png)

## Testing

- To test `GET /user` endpoint:
  1. Go to `https://bounswe2017group3.herokuapp.com/user` URL in your browser or send a `GET` request by a tool like Postman.  
  2. Server should respond with the `application/json;charset=UTF-8` content type. <img width="338" alt="screen shot 2017-04-16 at 00 33 04" src="https://cloud.githubusercontent.com/assets/13895224/25066915/83583d7c-223c-11e7-95f9-803ac82d37bd.png">  
  3. Server should render a JSON array which contains the current users in our database.  
- To test `POST /user` endpoint:
  1. Send a `POST` request to `https://bounswe2017group3.herokuapp.com/user` endpoint with a valid set of user parameters. An example query should look something like this: <img width="986" alt="screen shot 2017-04-16 at 00 35 23" src="https://cloud.githubusercontent.com/assets/13895224/25066950/4e6eb702-223d-11e7-9cf5-abed75cfe402.png">  
  2. Server should respond with the `application/json;charset=UTF-8` content type.  
  3. Server should not give any errors and render the newly created user object in the JSON format.
    - If you also want to test our database, send a `GET` request to `https://bounswe2017group3.herokuapp.com/user` endpoint after creating the user. It should render the newly created user in the array.
